VISION.md

```diff
@@ -0,0 +1,180 @@
+the most-specific configuration file. But that should be the decision of the
+developer, not of the configuration crate.
+
+<!-- TODO: more? -->
+
+## Users
+
+The config-rs-ng crate aims for the following users:
+
+* Small CLI App developers. A developer that implements a "small CLI tool", a
+  tool that might run for a few seconds and is non-interactive or has minimal
+  interactive features (it does not have to be explicitely stopped by the user
+  but runs to completion by itself). E.G.: `cat`, `toot`
+* Interactive CLI/TUI App developers: A developer that implements a CLI or TUI
+  that runs interactively and is a long-running process. The application
+  interacts with the user and waits for their input and is exited if the user
+  requests. E.G.: `less`, `vim`
+* GUI App developers: A developer that implements a graphical application. E.G.:
+  `libreoffice`, `ghostwriter`, `element` [^1]
+* Server App developers: A server binary that is started by some process manager
+  (`systemd`) and does interact with other software or Users via some remote
+  functionality but not interactively on CLI [^1]
+
+<!-- TODO: more? -->
+
+## Hard Requirements
+
+* Data sources
+  * A developer MUST be able to define own configuration data sources
+  * The crate MUST ship default configuration sources (From static string,
+    from file, from environment) Layering of configuration values from
+    different sources MUST be possible
+  * The crate MUST support layering of configuration sources, that is,
+    shadowning of configuration values by loading other configuration sources
+    and effectively "layer" them over the previous loaded configuration values.
+    * The crate MUST support "un-shadowing" configuration values if a "higher"
+      layer gets changed (via user interaction, config source changes or else)
+      and a value from a "lower" layer is no longer shadowed
+  * The crate MUST support loading configuration data sources in an synchronous
+    way as well as in an asynchronous way. The developer MUST be able to switch
+    loading mechanisms without having to rewrite their configuration types
+* The crate MUST allow deserializing the loaded configuration objects to a type
+  that implements `serde::Deserialize` although layering-information and context
+  is lost this way
+* The developer of an app MUST be able to find where a specific value of
+  configuration was defined
+* Configuration updates
+  * The developer of an app MUST be able to implement configuration updates
+    while their app runs, thus the crate MUST implement interfaces where the
+    developer can receive configuration updates.
+  * The crate MUST be able to "un-shadow" (read above) configuration values if
+    some configuration changes
+  * Still, the developer of an app SHOULD NOT be forced to use this
+    functionality.
+* Configuration file formats
+  * The crate SHOULD ship backends for the most common configuration file
+    formats (TOML, JSON, YAML)
```

Comment:
honestly, if I'm developing a CLI app and I only want to use a single config format (instead of letting users pick from a range of them), I definitely don't want to bloat my poor little binaries with crates for those formats, if I can help it. It would be great for _all_ of them to be feature-flagged (even if there's a set of them in default-features), so I can pick-and-choose.